New python manual (#372)
neo4j/driver-documentation: python-manual/modules/ROOT/pages/async.adoc

= Async query flow

> To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.

Reviewer: ```suggestion
To interact with the database in an asynchronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the synchronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.
```

> To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.

Reviewer 2: ```suggestion
To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.
```

`AsyncGraphDatabase` is not callable.

> To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define as `async` all functions that should be awaited.

Reviewer 2: ```suggestion
To interact with the database in an asyncronous way, create an `AsyncDriver` with `AsyncGraphDatabase()`. The workflow is very similar to the syncronous version, except that you must use `await` on all async calls, and define all functions that should be awaited  as `async`.
```

Author: I think it's clearer the way it is, so that `async` is as close to `define` as possible.

.An async driver example
[source,python, role=nocollapse]
----
import asyncio

> from neo4j import AsyncGraphDatabase

Reviewer 2: ```suggestion

from neo4j import AsyncGraphDatabase
```

:grimacing: the nitpicking... I can't help it, sorry.

async def get_people(tx):
    result = await tx.run("MATCH (a:Person) RETURN a.name AS name")
    records = await result.values()
    return records

async def main():
    URI = "<URI for Neo4j database>"
    AUTH = ("<Username>", "<Password>")

>     async with AsyncGraphDatabase.driver(URI, auth=AUTH) as driver:

Reviewer 2: ```suggestion
    uri = "<URI for Neo4j database>"
    auth = ("<Username>", "<Password>")

    async with AsyncGraphDatabase.driver(uri, auth=auth) as driver:
```
I'd only all-caps global constants.

Author: Moved out of main for consistency with other examples

        async with driver.session(database="neo4j") as session:
            records = await session.read_transaction(get_people)
            print(records)

asyncio.run(main())
----

[TIP]

> Async implements a concurrency model, but it is not the only possible one. Multithreading is also possible, although asyncio is often easier to implement in an application.

Reviewer: ```suggestion
Async implements a concurrency model, but it is not the only possible one. Multithreading is also possible, although `asyncio` is often easier to implement in an application.
```
